### PR TITLE
Make service service tests easier to use.

### PR DESCRIPTION
**Pull request by khalidharun**
https://www.pivotaltracker.com/story/show/101795046

### CODE REVIEW

File: R/servicetests.R [hunk 1]
```diff
@@ -17,22 +17,34 @@ get_cache_root <- function() { .servicetests_env$registry$.root }
...
 }
 
 
-#' Get the digest hash for the registered service tests.
-#'
-#' @export
-get_cache_digest <- function() { .servicetests_env$cache_digest }
+get_registered_keys <- function() {
+  reg <- get_registry()
+  lapply(reg, function(x) x$hash %||% generate_hash(c(x$fname, x$params)))
```
Comment (peterhurford) [inline R/servicetests.R]: Do you define `%||%` anywhere?  It's not in base R.

Comment (peterhurford) [inline R/servicetests.R]: Nice `lapply` by the way. :+1:

File: R/servicetests.R [hunk 2]
```diff
@@ -41,41 +53,65 @@ set_cache_root()
...
-  key <- generate_hash(append(params, fname, 0))
-
-  if (exists(key, .servicetests_env$reg)) stop("Service call already registered.")
+  hash <- generate_hash(c(params, fname))
+
+  reg <- .servicetests_env$reg
+
+  if (exists(key, reg)) {
+    if (identical(reg[[key]]$hash, hash)) {
+      if (!force) {
```
Comment (peterhurford) [inline R/servicetests.R]: Use `identical(force, FALSE)`

File: R/servicetests.R [hunk 3]
```diff
@@ -41,41 +53,65 @@ set_cache_root()
...
+        warning(pp("Service call is already registered. Set ", crayon::cyan("force"), " parameter to ", crayon::cyan("TRUE"), " to re-register call if the response has been updated."))
```
Comment (peterhurford) [inline R/servicetests.R]: Does `pp` do this sort of thing?  Or do you mean `paste`?

Personally I'd prefer replacing all instances of `pp` with `paste`, especially in packages where it removes you a dependency.

Also you should prefix it `productivus::pp` and include it in your `Imports` if you really want it (not recommended).

File: R/servicetests.R [hunk 4]
```diff
@@ -41,41 +53,65 @@ set_cache_root()
...
+        return()
```
Comment (peterhurford) [inline R/servicetests.R]: Should you `return` with a value?

File: R/servicetests.R (hunk 3 above)
Comment (khalidharun) [inline R/servicetests.R]: good point. removing pp and crayon dependency.

**Comment by khalidharun** [comment]
Thanks for the review @peterhurford I addressed them all.

File: R/servicetests.R [hunk 5]
```diff
@@ -16,66 +16,96 @@ set_cache_root <- function(cache_root = getOption("servicetests.dir", "~/.R/serv
...
+  hash <- generate_hash(c(params, fname))
 
-  if (exists(key, .servicetests_env$reg)) stop("Service call already registered.")
+  reg <- .servicetests_env$reg
 
-  response <- eval(sfn, envir)
-  .servicetests_env$reg[[key]] <- list(fname = fname, params = params, response = response)
+  if (exists(key, reg)) {
+    if (identical(force, FALSE)) {
+      warning(paste0("Service call is already registered. Set 'force = TRUE' to replace service call."))
```
Comment (peterhurford) [inline R/servicetests.R]: What is the `paste0` for?

File: R/servicetests.R [hunk 6]
```diff
@@ -88,6 +118,13 @@ test_service <- function(obj, envir = parent.frame(), on.pass_each = NULL, on.fa
   if (!is.null(on.fail_each) && !is.function(on.fail_each)) stop("on.fail_each must be a function.")
   if (!is.null(on.pass_each) && !is.function(on.pass_each)) stop("on.pass_each must be a function.")
 
+  all.equal.data.frame <- function (target, current, ...)
+  {
+    rownames(current) <- NULL
+    rownames(target) <- NULL
+    all.equal.default(target, current, ...)
+  }
```
Comment (peterhurford) [inline R/servicetests.R]: Nice. :+1:

File: R/servicetests.R [hunk 7]
```diff
@@ -105,28 +142,34 @@ test_service <- function(obj, envir = parent.frame(), on.pass_each = NULL, on.fa
...
 #' @export
-test_all_services <- function(return.type = "verbose", on.pass_all = NULL, on.fail_any = NULL, ...) {
+test_service_calls <- function(keys = NULL, return.type = "verbose", on.pass_all = NULL, on.fail_any = NULL, ...) {
   return.type <- match.arg(return.type, c("verbose", "logical"))
   if (!is.null(on.pass_all) && !is.function(on.pass_all)) stop("on.pass_all must be a function.")
   if (!is.null(on.fail_any) && !is.function(on.fail_any)) stop("on.fail_any must be a function.")
+  reg <- get_registry()
+  if (!missing(keys) && !is.null(keys)) {
+    reg <- reg[which(names(reg) %in% keys)]
+    if (!length(reg)) stop("Key not found")
```
Comment (peterhurford) [inline R/servicetests.R]: I think `length(reg) == 0` is more clear.

File: R/servicetests.R [hunk 8]
```diff
@@ -16,66 +16,96 @@ set_cache_root <- function(cache_root = getOption("servicetests.dir", "~/.R/serv
...
+  })
+
+  .servicetests_env$reg[[key]] <- list(key = key, fname = fname, params = params, hash = hash, response = response)
+  TRUE
 }
 
 #' register a list of service calls
 #' @param call_list lanuage. The service calls to register.
 #' @param envir environment. The environment where the service call is evaluated.
-#' @export
```
Comment (peterhurford) [inline R/servicetests.R]: Are you no longer exporting this function?

File: R/servicetests.R [hunk 9]
```diff
@@ -88,6 +118,13 @@ test_service <- function(obj, envir = parent.frame(), on.pass_each = NULL, on.fa
   if (!is.null(on.fail_each) && !is.function(on.fail_each)) stop("on.fail_each must be a function.")
   if (!is.null(on.pass_each) && !is.function(on.pass_each)) stop("on.pass_each must be a function.")
 
+  all.equal.data.frame <- function (target, current, ...)
```
Comment (peterhurford) [inline R/servicetests.R]: You may need to document this.

File: R/servicetests.R (hunk 8 above)
Comment (khalidharun) [inline R/servicetests.R]: yeah, no longer exporting it.

**Pull request merged**